inlanecruising_plugin/src/inlanecruising_plugin.cpp

```diff
@@ -392,6 +397,62 @@ std::vector<cav_msgs::TrajectoryPlanPoint> InLaneCruisingPlugin::compose_traject
   return traj_points;
 }
 
+double InLaneCruisingPlugin::get_adaptive_lookahead(double velocity){
+  
+  // lookahead:
+  // v<10kph:  5m
+  // 10kph<v<50kph:  0.5*v
+  // v>50kph:  25m
+
+  double lookahead = config_.minimum_lookahead_distance;
+
+  if (velocity < config_.minimum_lookahead_speed)
+  {
+    lookahead = config_.minimum_lookahead_distance;
+  } 
+  else if (velocity >= config_.minimum_lookahead_speed && velocity < config_.maximum_lookahead_speed)
+  {
+    lookahead = config_.lookahead_ratio * velocity;
+  } 
+  else lookahead = config_.maximum_lookahead_distance;
+
+  return lookahead;
+
+}
+
+std::vector<double> InLaneCruisingPlugin::get_lookahead_speed(const std::vector<lanelet::BasicPoint2d>& points, const std::vector<double>& speeds, const double& lookahead){
+  
+  if (lookahead < config_.minimum_lookahead_distance)
+  {
+    throw std::invalid_argument("Invalid lookahead value");
+  }
+
+
+  if (speeds.size() != points.size())
+  {
+    throw std::invalid_argument("Speeds and Points lists not same size");
+  }
+
+  std::vector<double> out;
+  out.reserve(speeds.size());
+
+  for (int i = 0; i < points.size(); i++)
+  {
+    int idx = i;
+    double min_dist = std::numeric_limits<double>::max();
+    for (int j=i+1; j < points.size(); j++){
+      double dist = lanelet::geometry::distance2d(points[i],points[j]);
+      if (abs(lookahead - dist) <= min_dist){
```

Comment:
Seems sensible to me. Could produce weirdness if we have some kind of ultra tight u-turn or something, where there's a point at a higher index that's actually closer to being lookahead dist from our current point, but that feels like an extremely esoteric case and such a trajectory should never exist so it doesn't really give me cause for concern.